addons/product_margin/models/product_product.py

```diff
@@ -128,38 +125,54 @@ def _compute_product_margin_fields_values(self, field_names=None):
-            ctx = self.env.context.copy()
-            ctx['force_company'] = company_id
-            invoice_types = ('in_invoice', 'in_refund')
-            self.env.cr.execute(sqlstr, (val.id, states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
-            result = self.env.cr.fetchall()[0]
-            res[val.id]['purchase_avg_price'] = result[0] and result[0] or 0.0
-            res[val.id]['purchase_num_invoiced'] = result[1] and result[1] or 0.0
-            res[val.id]['total_cost'] = result[2] and result[2] or 0.0
-            res[val.id]['normal_cost'] = val.standard_price * res[val.id]['purchase_num_invoiced']
-            res[val.id]['purchase_gap'] = res[val.id]['normal_cost'] - res[val.id]['total_cost']
+        invoice_types = ('out_invoice', 'out_refund')
+        self.env.cr.execute(sqlstr, (tuple(self.ids), states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
+        products_vals = {
+            product_id: {'avg_unit_price': avg, 'num_qty': qty, 'total': total, 'sale_expected': sale}
+            for product_id, avg, qty, total, sale in self.env.cr.fetchall()
+        }
+        for product_id in self.ids:
+            avg, qty, total, sale = None, None, None, None
+            if products_vals.get(product_id):
+                product_vals = products_vals[product_id]
+                avg, qty, total, sale = product_vals['avg_unit_price'], product_vals['num_qty'], product_vals['total'], product_vals['sale_expected']
+            res[product_id]['sale_avg_price'] = avg and avg or 0.0
+            res[product_id]['sale_num_invoiced'] = qty and qty or 0.0
+            res[product_id]['turnover'] = total and total or 0.0
+            res[product_id]['sale_expected'] = sale and sale or 0.0
+            res[product_id]['sales_gap'] = res[product_id]['sale_expected'] - res[product_id]['turnover']
 
-            res[val.id]['total_margin'] = res[val.id]['turnover'] - res[val.id]['total_cost']
-            res[val.id]['expected_margin'] = res[val.id]['sale_expected'] - res[val.id]['normal_cost']
-            res[val.id]['total_margin_rate'] = res[val.id]['turnover'] and res[val.id]['total_margin'] * 100 / res[val.id]['turnover'] or 0.0
-            res[val.id]['expected_margin_rate'] = res[val.id]['sale_expected'] and res[val.id]['expected_margin'] * 100 / res[val.id]['sale_expected'] or 0.0
-            for k, v in res[val.id].items():
-                setattr(val, k, v)
+        ctx = self.env.context.copy()
+        ctx['force_company'] = company_id
+        invoice_types = ('in_invoice', 'in_refund')
+        self.env.cr.execute(sqlstr, (tuple(self.ids), states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
+        products_vals = {
+            product_id: {'avg_unit_price': avg, 'num_qty': qty, 'total': total}
+            for product_id, avg, qty, total, _ in self.env.cr.fetchall()
```

Comment:
```suggestion
            for product_id, avg, qty, total, dummy in self.env.cr.fetchall()
```
less error-prone, in case of a future change while introducing a translatable string